### ISSUE TITLE
Modal dialogs

### ISSUE THREAD

**Issue by loicsans** [akvo/akvo-lumen#818]
**Item referencing on modals**

# Context
In order to consistently handle item (dataset, visualization, collection, user, ..) referencing in modals and other places

# Problem
Currently we have used a number of conventions as in the following examples:
- Delete dataset "Youth tobbacco use"
- Delete collection Some collection name

# Proposal
Use a colon to reference items e.g.
- Share visualization: Visualization name
- Delete dataset: Dataset name
- Delete user: Firstname Lastname

![image](https://cloud.githubusercontent.com/assets/12692444/26586244/51421842-4557-11e7-8103-d2dd8fbe4ad9.png)

**Share asset modal**

# Problem
We need to place elements in an order and simplify the modal itself

<img width="511" alt="screen shot 2017-05-30 at 15 36 40" src="https://cloud.githubusercontent.com/assets/12692444/26583435/e8843564-454d-11e7-9aee-ec543dcfd285.png">

Open:
- How to share multiple assets:
- Is the modal different when sharing from within an asset screen or from the lib/collection?

# Proposal
Could we kill a number of birds in presenting it as follows:
 
![image](https://cloud.githubusercontent.com/assets/12692444/26585362/8b2797d8-4554-11e7-863b-366d83d30a2c.png)

**New collection modal improvements**

# Idea
<img width="508" alt="screen shot 2017-05-30 at 13 34 18" src="https://cloud.githubusercontent.com/assets/12692444/26579493/c578a7fa-453c-11e7-9f82-9cda28a155cc.png">

# Proposal
- minimize header; New collection suffices
- provide similar naming convention as other items; Untitled collection, as the default
- place Cancel and Create in proximity

![image](https://cloud.githubusercontent.com/assets/12692444/26579814/2a3ff35e-453e-11e7-80e1-2ebb7d51fcdf.png)

**Modal dialog dimensions**

# Context
[Borrowing descriptions](https://material.io/guidelines/components/dialogs.html#) Dialog types include:

- Alerts are urgent interruptions that inform about a situation and require acknowledgement.
- Simple menus display options for list items, whereas simple dialogs can provide details or actions about a list item.
- Confirmation dialogs require users to explicitly confirm a choice.
-- [Dialog usage guidelines](https://www.nngroup.com/articles/modal-nonmodal-dialog/)
# Problem
In this example, the dialog is extra long that is required by its content; which leave the user in confusion as unto whether some content is missing in the middle

<img width="514" alt="screen shot 2017-05-30 at 11 13 53" src="https://cloud.githubusercontent.com/assets/12692444/26574196/1a43474a-4529-11e7-99ef-09b2ecbc49fd.png">

# Proposal
The height and width dimensions need to defined; 
- the height by content 
- the width could be fixed for all modal dialogs

- About the placement: I am thinking centred dialogs appear 'well' anchored to the page in contrast to those hanging at the top, opinions?   

Dimensions to cater for:

![image](https://cloud.githubusercontent.com/assets/12692444/26578149/6a95a5cc-4537-11e7-98fb-075f6e4f3654.png)

- Furthermore, the buttons could be placed in close proximity to each other as illustrated, allowing smoother reading/scanning of possible actions, i.e. the user doesnt need to swing their eyes left to right to view the actions

The end result should be simple e.g. as shown below, to ensure that it takes as minimal time as possible for the user to accomplish their task or dismiss it

![image](https://cloud.githubusercontent.com/assets/12692444/26578565/085bd7e4-4539-11e7-86f4-abf0095aa145.png)

# For Modals with lots of content: we need to show at least 2 items all the time:
(i) Modal title, (ii) Description, (iii) Actions: to ensure the user is able to determine the course of action, access content/buttons without having to scroll these 3 out of view e.g.

![image](https://cloud.githubusercontent.com/assets/12692444/26586978/a623fba8-4559-11e7-8285-0158f5ee4760.png)

**Unify dialogue boxes**

# Context
Review dialog boxes, error messages, exceptions, ..

# Problem or idea
We need to unify the 'language', descriptions, buttons or Actions e.g. the dialog below shd have direct title, text and actions

<img width="611" alt="screen shot 2017-04-27 at 10 39 07" src="https://cloud.githubusercontent.com/assets/12692444/25472868/d574ee0a-2b35-11e7-8767-7851bef431e2.png">

# Solution or next step
Example
- Title: Remove user "Nadia" - we shd user Name in place of Email here.. 
- Description: no need for a desc if the action is straight forward
- Actions: Default | Primary: "Cancel | Remove"

**Refine New dataset Dialog**

# Problem or idea
There is a number of improvements to make here so i will just get to what we might need to do

<img width="478" alt="screen shot 2017-05-29 at 16 47 16" src="https://cloud.githubusercontent.com/assets/12692444/26552129/83840d02-448e-11e7-8ae4-8f4ba838a81a.png">

# Solution or next step
- Full-screen-size the process; it is not necessary to have this as a pop-up
- Place the Checkbox on the lefthand side of the selection "File has column headers". On this option, we seem to have deviated from the workflow we had (if i remember correctly, it was in the last screen, where a preview would be available so the user didnt have to think or remember or go open the file to know for sure if there is a header and so on..) - please confirm. 
- The text No file selected is not necessary - I think the action to be taken shd be easy to determine given the two options: (i) Drop file [...] OR (ii) Select file
- As in the naming convention that we have with Visualizations, shd we provide the 'naming' for the new dataset at the top? > Change it to "Import Dataset"
- Indicate process/progress - atm this look like tabs and remotely communicate progress. we could do it as follows:

![image](https://cloud.githubusercontent.com/assets/12692444/26553212/57b562a2-4493-11e7-9941-b8035d12a55e.png)

**Comment by loicsans** [comment]
Priority: 10
Effort: 6

**Comment by loicsans** [comment]
More on closing / dismissing modal dialogs. along with the Cancel button, Close button (cross) and Escape key...Clicking outside the window would be good.

**Comment by gabemart** [comment]
The technical background work on this has taken much longer than anticipated because modals were implemented in several different ways. I've made some progress in unifying them and extracting common components, but modals are still divided between those that use Redux actions for display/hide and callbacks, and those that are injected like regular components. 

After discussion with @jonase we've decided it would probably be best to not use any redux actions for the standard modal workflow - it complicates things a lot and offers few benefits in practise.

That's out of scope for this milestone though. So I'll commit the unification work so far, which should be enough to allow consistent styling and some of the styling and ux improvements given in this issue.
